### PR TITLE
Replace String comparisons with "", String() to is_empty()

### PR DESCRIPTION
**Pull request by nathanfranke**
Regex to replace:
```
((?:[a-zA-Z0-9_]+::)?([a-zA-Z0-9_]+(\(\))?(\[.+\])*)((\.|->)(([a-zA-Z0-9_]+(\(\))?(\[.+\])*)))*) == (String\(\)|"")
```
to
```
$1.is_empty()
```
---
```
((?:[a-zA-Z0-9_]+::)?([a-zA-Z0-9_]+(\(\))?(\[.+\])*)((\.|->)(([a-zA-Z0-9_]+(\(\))?(\[.+\])*)))*) != (String\(\)|"")
```
to
```
!$1.is_empty()
```

Include files:
```
*.cpp, *.h, *.hpp
```
Ignore files:
```
thirdparty/*, tests/core/test_string.h
```

~~Opening this as a draft to encourage discussion on how this condition should be formatted. Also, possible recommendations for the format script/check.~~

**BENCHMARKS** (Thanks @Xrayez)
```
===============================================================================
tests/test_string.h:1300:
TEST CASE:  [Stress][String] Empty via ' == String()'

0.017540 s: [Stress][String] Empty via ' == String()'
===============================================================================
tests/test_string.h:1309:
TEST CASE:  [Stress][String] Empty via `is_empty()`

0.015278 s: [Stress][String] Empty via `is_empty()`
===============================================================================
```

---

# Important: This PR also:

~~- Adds `is_empty` and `length` to `string_name.h`~~
- Adds two stress tests to `test_string.h`
- Changes to `.empty()` on `std::string`s
~~- Makes some formatting/organization on `string_name.h`~~

All manual changes are in my branch, [`string-empty-manual`](https://github.com/nathanfranke/godot/tree/string-empty-prelude)

### CODE REVIEW

**Comment by nathanfranke** [comment]
Oops, accidentally made 413 changes without once trying to compile. Working on a fix. :facepalm:

**Comment by Xrayez** [comment]
You can try write performance tests for `== String()` vs `empty()` as well (without making assertions), something like:
```c++
TEST_CASE("[Stress][String] Empty via ' == String()'") {
    for (int i = 0; i < 100000; ++i) {
        String str = "Hello World!";
        if (str == String()) {
            continue;
        }
	}
}

TEST_CASE("[Stress][String] Empty via `empty()`") {
    for (int i = 0; i < 100000; ++i) {
        String str = "Hello World!";
        if (str.empty()) {
            continue;
        }
	}
}
```
and then run:
```
.\bin\godot --test --source-file="*test_string*" --test-case="*Stress*" --duration
```

See [Unit testing](https://docs.godotengine.org/en/latest/development/cpp/unit_testing.html).

**Comment by Anutrix** [comment]
There's a lot of place where `== ""` or `!= ""` is used. Maybe that can be replaced too.

**Comment by nathanfranke** [comment]
Q: Should `StringName` contain an `empty` method?

**Comment by nathanfranke** [comment]
> There's a lot of places where `== ""` or `!= ""` are used. Maybe those can also be replaced too.

Yes, indeed. Thanks for the notice. The lines changed have doubled!

**Comment by nathanfranke** [comment]
> Q: Should `StringName` contain an `empty` method?

I'm going to say yes.

---

Instances of `if (s.length())` should also be replaced. Performance unsure.

**Comment by aaronfranke** [comment]
@nathanfranke This likely needs to be redone from scratch. Godot now uses `is_empty` instead of `empty`.

**Comment by akien-mga** [comment]
Welp, we have a lot of teams :P

Screenshot_20210410_142605

It would be good to amend the commit message to be more explicit (e.g. match the PR's title, which is good).

**Comment by nathanfranke** [comment]
> Welp, we have a lot of teams :P
> 
> Screenshot_20210410_142605
> 
> It would be good to amend the commit message to be more explicit (e.g. match the PR's title, which is good).

Done, and I hope I can get these bugs ironed out since I feel bad for spamming GitHub actions- though I'm on Linux so I can't try compiling for iOS, OSX, Windows... ☹

**Comment by akien-mga** [comment]
Note that the compilation issues show that there's bugs with the regex used. IMO to ensure that the final set of changes is safe, you need to tweak the regex until they produce a consistent result when applied from scratch (i.e. not with manual fixes of the few places where compilation fails). Otherwise we wouldn't be confident that we're not introducing invalid changes which just happen to compile fine.

**Comment by nathanfranke** [comment]
> Note that the compilation issues show that there's bugs with the regex used. IMO to ensure that the final set of changes is safe, you need to tweak the regex until they produce a consistent result when applied from scratch (i.e. not with manual fixes of the few places where compilation fails). Otherwise we wouldn't be confident that we're not introducing invalid changes which just happen to compile fine.

This might be my laziness speaking, but to make the regex truly "bug-less" it would basically have to be a C++ syntax parser and understand everything like order of operations. At that point it would be more likely for there to be a bug with the regex than a non-error bug with the current regex.

File: tests/test_string.h
Comment (aaronfranke) [inline tests/test_string.h:1339]: I think you accidentally ran your regex on this test case.

**Review by aaronfranke — CHANGES_REQUESTED** [review]

File: core/string/string_name.h
Comment (neikeq) [inline core/string/string_name.h:128]: `operator String()` can be an expensive operation. I'm not sure the impact when the data is stored as a `String`, it may not be a problem there. But when the data is a static C string, `operator String()` will create a new `String` from that every time. I don't hiding something like that into a seemingly quick function like `length()`.

File: core/string/string_name.h
Comment (neikeq) [inline core/string/string_name.h:131]: All empty checks I've seen before were `my_string_name == StringName()` which is pretty much `my_string_name._data == nullptr`. I just checked the `StringName` constructors and can confirm an empty string is never allowed to be stored, so a null check is enough (specially important if `length()` can be slow).

**Review by neikeq — DISMISSED** [review]

File: core/string/string_name.h
Comment (nathanfranke) [reply core/string/string_name.h:131]: Thank you very much for this. It also lets me remove the above String cast which I was really uncomfortable with as well.

**Comment by nathanfranke** [comment]
@aaronfranke @neikeq How does this look for `length` and `is_empty` check for `StringName`? I know it's verbose but this is both memory and speed efficient, especially where it matters.

```c++

	_FORCE_INLINE_ int length() const {
		if (_data) {
			if (_data->cname) {
				int len = 0;
				const char *ptr = _data->cname;
				while (*(ptr++) != 0) {
					len++;
				}
				return len;
			} else {
				return _data->name.length();
			}
		} else {
			return 0;
		}
	}

	_FORCE_INLINE_ bool is_empty() const { return _data == nullptr; }

```

**Review by aaronfranke — APPROVED** [review]
I haven't looked over every single line (since there's so many), but I looked over a few dozen files and the important ones like StringName and the tests. It looks good to me.

**Comment by akien-mga** [comment]
I know it's a hassle, but I think it would be better if this PR was split into 3:
- Changes to use `is_empty()` for `String` (can be merged right away, should not have any impact on behavior)
- New methods added to `StringName` (needs review from Core team)
- Changes to use new methods for `StringName` (once the previous one has been approved and merged)

**Comment by nathanfranke** [comment]
@akien-mga Done, this pull request is the first bullet point you requested.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by reduz** [comment]
Time passes and I am not very happy about this as-is, I think it makes code more difficult to read. If you have something:

```c++
if (very_long_piece_of_text != "")
// and you replace by 
if (!very_long_piece_of_text.is_empty())
```
In most cases readability is very affected and I find missing or forgetting a `!` is quite a common cause of bugs. This is why in classes like `RID()` and `Ref<>` have methods `is_null()` and `is_valid()`. 

I am not particularly against doing this change, but I think we should have `is_not_empty()`, so code like:
```c++
if (!very_long_piece_of_text.is_empty())
// Should be instead
if (very_long_piece_of_text.is_not_empty())
```

**Pull request merged**
